contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,309 @@
+    if (!fs.exists(parentPath)) {
+      throw new FileNotFoundException(s"No such file or directory: $parentPath")
+    }
+
+    super.listFrom(path, hadoopConf)
+  }
+
+  /**
+   * Write file with actions under a specific path.
+   */
+  protected def writeActions(
+      fs: FileSystem,
+      path: Path,
+      actions: Iterator[String]
+  ): Unit = {
+    logDebug(s"writeActions to: $path")
+    val stream = fs.create(path, true)
+    actions.map(_ + "\n").map(_.getBytes(UTF_8)).foreach(stream.write)
+    stream.close()
+  }
+
+  def write(
+      path: Path,
+      actions: Iterator[String],
+      overwrite: Boolean = false
+  ): Unit = {
+    write(path, actions, overwrite, getHadoopConfiguration)
+  }
+
+  override def write(
+      path: Path,
+      actions: Iterator[String],
+      overwrite: Boolean,
+      hadoopConf: Configuration
+  ): Unit = {
+    val (fs, resolvedPath) = resolvePath(path, hadoopConf)
+
+    if (overwrite) {
+      return writeActions(fs, path, actions);
+    };
+
+    val parentPath = resolvedPath.getParent
+    assert(FileNames.isDeltaFile(path))
+    val version = FileNames.deltaVersion(path)
+
+    if (version > 0) {
+      val prevVersion = version - 1
+      val prevPath = FileNames.deltaFile(parentPath, prevVersion)
+      getDbEntry(prevPath) match {
+        case Some(entry) => fixDeltaLog(fs, entry)
+        case None =>
+          if (!fs.exists(prevPath)) {
+            throw new java.nio.file.FileSystemException(
+              s"previous commit ${prevPath} doesn't exist"
+            )
+          }
+          // previous commit exists in fs but not in dynamodb
+      }
+    } else {
+      getDbEntry(path) match {
+        case Some(entry) =>
+          if (entry.complete && !fs.exists(path)) {
+            throw new java.nio.file.FileSystemException(
+              s"Old entries for ${parentPath} still exist in the database"
+            )
+          }
+        case None => ;
+      }
+    }
+
+    val tempPath = createTemporaryPath(resolvedPath)
+
+    val entry =
+      LogEntry(
+        resolvedPath.getParent,
+        resolvedPath.getName,
+        tempPath,
+        false,
+        None
+      )
+
+    logDebug(s"write: writing to temporary file ${tempPath}")
+    writeActions(fs, entry.absoluteTempPath(), actions);
+
+    // commit to dynamodb E(N, false) and exclude commitTime
+    logDebug(s"write incomplete entry to dynamodb ${entry}")
+    putDbEntry(entry, overwrite = false)
+
+    try {
+      onWriteCopyTempFile();
+      copyFile(fs, entry.absoluteTempPath(), resolvedPath)
+      // Commit (with overwrite=true) to DynamoDB entry E(N, true)
+      // with commitTime attribute in epoch seconds
+      onWritePutDbEntry();
+      putDbEntry(entry.asComplete(), overwrite = true)
+    } catch {
+      case e: Throwable =>
+        logInfo(s"${e.getClass.getName}: ignoring recoverable error: $e")
+    }
+  }
+
+  protected def onWriteCopyTempFile(): Unit = {}
```

Comment:
I'm a little worried about child implementations doing something unexpected here.

I'd suggest either:
- A) we rename these methods to indicate that they should only be used for error injection, e.g. `injectOnWriteCopyTempFileError`
- or
- B) these methods take params, and we put meaningful implementations into them. e.g.
```scala
protected def onWriteCopyTempFile(fs: ..., entry: ..., resolvedPath: ...): Unit = {
  copyFile(fs, entry.absoluteTempPath(), resolvedPath)
}
```

I prefer option B. what do you think?

Reply:
cc @zsxwing

Reply:
I was thinking about B too, implemented this way.